chore: add versioning for drop notify messages

sypakine opened the pull request:
Related: #35057

Aligns the control-plane drop notify struct with the dataplane struct, which contains a version field. This is modeled on a similar change to trace notify, which is versioned.

Also changed `__notify_pktcap_hdr` function to accept a version, as the version for each notification can evolve independently.

Similar change was done in https://github.com/cilium/cilium/pull/9321

---

When making changes to the datapath structs produced, we need to be congnizant of how these messages are consumed. Below considers the most interesting/problematic cases and their result:

1. For the upgrade case:
- dataplane format is the same
- cap field value is handled correctly in control-plane

2. For Downgrade case:
- dataplane format is the same
- cap field value is handled as it currently is

Future version iterations to consider downgrade cases.

```release-note
Add versioning to drop notify events.
```

gandro reviewed (commented):
Hubble-wise this looks okay. I'm not sure I fully grasp the consequences of putting the version in the old length field though. What happens a user downgrades from Cilium v1.17 (with this patch applied) to v1.16 (without this patch applied). 

Concretely I'm wondering if we have code that interprets the old `DropNotify.CapLen` and gets an invalid value (because it interprets the version as part of the length). From a quick glance don't see any code actually using `CapLen` (besides checking that it's larger than 0), so it seems fine - but I think it would be nice to have it explained in the commit message why this change is backwards-compatible

gandro reviewed (approved):
Awesome, thank you for enumerating the different upgrade/downgrade cases. LGTM!

kaworu commented on pkg/monitor/datapath_drop.go:
```diff
@@ -15,8 +15,16 @@ import (
 const (
-	// DropNotifyLen is the amount of packet data provided in a drop notification
-	DropNotifyLen = 36
+	DropNotifyVersion1 = 1 << iota
```
I don't understand why it's defined that way :thinking: because following this logic:
```go
DropNotifyVersion2 = 1 << iota
DropNotifyVersion3 = 1 << iota
```
And we end up with `DropNotifyVersion3=4`.

sypakine replied:
Only the first entry needs to use `... = 1 <<` as it sets the initial index.

Subsequent version would be defined as:
```
DropNotifyVersion1 = 1 << iota
DropNotifyVersion2
// const DropNotifyVersion2 untyped int = 2
```

kaworu replied:
Right thanks for pointing it out, I wrote the incorrect definitions to explicit my point: that we'll eventually end up with `DropNotifyVersion3 = 4` and I don't understand why we wouldn't do `DropNotifyVersion3 = 3` instead.

sypakine replied:
Changed to `DropNotifyVersion1 = 1`

kaworu commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -108,7 +108,7 @@ func (p *Parser) Decode(data []byte, decoded *pb.Flow) error {
 	switch eventType {
 	case monitorAPI.MessageTypeDrop:
-		packetOffset = monitor.DropNotifyLen
+		packetOffset = monitor.DropNotifyV1Len
```
Since this patch introduces `dn.DataOffset()`, should we use it after decoding here instead? See how it is implemented for the following `monitor.TraceNotify` `case`.

sypakine replied:
Done, thanks.

kaworu commented on bpf/lib/policy_log.h:
```diff
@@ -94,7 +94,7 @@ send_policy_verdict_notify(struct __ctx_buff *ctx, __u32 remote_label, __u16 dst
 	msg = (typeof(msg)) {
 		__notify_common_hdr(CILIUM_NOTIFY_POLICY_VERDICT, 0),
-		__notify_pktcap_hdr(ctx_len, (__u16)cap_len),
+		__notify_pktcap_hdr(ctx_len, (__u16)cap_len, NOTIFY_CAPTURE_VER),
```
ditto new constant

sypakine replied:
See comment about `NOTIFY_CAPTURE_VER`

kaworu commented on bpf/lib/dbg.h:
```diff
@@ -220,7 +220,7 @@ static __always_inline void cilium_dbg_capture2(struct __ctx_buff *ctx, __u8 typ
 	struct debug_capture_msg msg = {
 		__notify_common_hdr(CILIUM_NOTIFY_DBG_CAPTURE, type),
-		__notify_pktcap_hdr(ctx_len, (__u16)cap_len),
+		__notify_pktcap_hdr(ctx_len, (__u16)cap_len, NOTIFY_CAPTURE_VER),
```
Should we add a use a different constant here than re-using the one for notify messages?
```suggestion
		__notify_pktcap_hdr(ctx_len, (__u16)cap_len, DBG_VER),
```

sypakine replied:
My reading is this is already generic for any notify message (trace, policy, debug, and drop).

Expanded on the comment to clarify this point.

kaworu replied:
Sure, but if the same constant is used for all the message types I don't see how they can evolve independently as the commit message suggest. Not blocking though, extracting it can pave the way for future changes.

sypakine replied:
future commits would create a new constant: e.g. `bpf/lib/drop.h`

```
#define NOTIFY_DROP_VER = 2;
...

	msg = (typeof(msg)) {
		__notify_common_hdr(CILIUM_NOTIFY_DROP, (__u8)error),
		__notify_pktcap_hdr(ctx_len, (__u16)cap_len, NOTIFY_DROP_VER),
		...
```

kaworu commented on bpf/lib/drop.h:
```diff
@@ -77,7 +77,7 @@ int __send_drop_notify(struct __ctx_buff *ctx)
 	msg = (typeof(msg)) {
 		__notify_common_hdr(CILIUM_NOTIFY_DROP, (__u8)error),
-		__notify_pktcap_hdr(ctx_len, (__u16)cap_len),
+		__notify_pktcap_hdr(ctx_len, (__u16)cap_len, NOTIFY_CAPTURE_VER),
```
ditto new constant

sypakine replied:
See comment about `NOTIFY_CAPTURE_VER`

kaworu reviewed (changes requested):
Thanks for the PR @sypakine! Some questions and suggestions inline.

Also I'm not sure to understand the upgrade/downgrade rational described [here](https://github.com/cilium/cilium/pull/35413#issuecomment-2451890790). You mention unchanged zero values for `version` but `NOTIFY_CAPTURE_VER=1` both before and after this patch. Here is my understanding:

1. the patch doesn't change the datapath behaviour, the version has been extracted from the `__notify_pktcap_hdr` macro but we still consistently provide the same value. Thus both before and after the patch we get `version=NOTIFY_CAPTURE_VER` for all messages types. Consequently, I don't see any upgrade/downgrade concern.
2. The user space interpretation of `CapLen` is now different as we use the first 16 bits following `OrigLen` to extract it (instead of the first 32 bits before this patch), the following 16 bits being the `version` (both in datapath and user space after this patch). This should cause no issue because every usages of `CapLen` is also checking the payload length (see [here](https://github.com/cilium/cilium/blob/c345a69888090f8f3bb8ddde2fe66b4c8914f341/pkg/monitor/datapath_drop.go#L103) and [here](https://github.com/cilium/cilium/blob/c345a69888090f8f3bb8ddde2fe66b4c8914f341/pkg/monitor/datapath_drop.go#L113)). So arguably we don't need `CapLen` at all, since we can always compute it from `len(payload.data) - monitor.DropNotifyV1Len`.

Does that make sense or am I missing something?

kaworu reviewed (approved):
Thanks @sypakine !

rolinh reviewed (approved):
Thanks, lgtm.
@sypakine it seems that there were some temporary CI issues. Try rebasing before re-triggering the tests.

sypakine committed:
chore: add versioning for drop notify messages

Aligns the control-plane drop notify struct with the dataplane struct, which contains a version field. This is modeled on a similar change to trace notify, which is versioned.

Also changed __notify_pktcap_hdr function to accept a version, as the version for each notification can evolve independently.

Similar change was done in https://github.com/cilium/cilium/pull/9321

---

When making changes to the datapath structs produced, we need to be congnizant of how these messages are consumed. Below considers the most interesting/problematic cases and their result:

1. For the upgrade case:
- dataplane format is the same
- cap field value is handled correctly in control-plane

2. For Downgrade case:
- dataplane format is the same
- cap field value is handled as it currently is

Future version iterations to consider downgrade cases.

---

Signed-off-by: Mark St John <markstjohn@google.com>
